core/crdt/pncounter.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2022 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package crdt
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/fxamacker/cbor/v2"
+	dag "github.com/ipfs/boxo/ipld/merkledag"
+	ds "github.com/ipfs/go-datastore"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/ugorji/go/codec"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/base"
+	"github.com/sourcenetwork/defradb/errors"
+)
+
+var (
+	// ensure types implements core interfaces
+	_ core.ReplicatedData = (*PNCounterRegister)(nil)
+	_ core.Delta          = (*PNCounterDelta)(nil)
+)
+
+// PNCounterDelta is a single delta operation for an PNCounterRegister
+type PNCounterDelta struct {
+	DocKey    []byte
+	FieldName string
+	Priority  uint64
+	// SchemaVersionID is the schema version datastore key at the time of commit.
+	//
+	// It can be used to identify the collection datastructure state at the time of commit.
+	SchemaVersionID string
+	Data            []byte
+	FieldValue      client.FieldValue `json:"-"` // should not be marshalled
+}
+
+// GetPriority gets the current priority for this delta.
+func (delta *PNCounterDelta) GetPriority() uint64 {
+	return delta.Priority
+}
+
+// SetPriority will set the priority for this delta.
+func (delta *PNCounterDelta) SetPriority(prio uint64) {
+	delta.Priority = prio
+}
+
+// Marshal encodes the delta using CBOR.
+func (delta *PNCounterDelta) Marshal() ([]byte, error) {
+	b, err := delta.FieldValue.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	delta.Data = b
+
+	h := &codec.CborHandle{}
+	buf := bytes.NewBuffer(nil)
+	enc := codec.NewEncoder(buf, h)
+	err = enc.Encode(delta)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// PNCounterRegister, Last-Writer-Wins Register, is a simple CRDT type that allows set/get
+// of an arbitrary data type that ensures convergence.
+type PNCounterRegister struct {
```

Comment:
A register is a type that can only hold a single value, of which the only exposed mutator of that type is "Set". The `LWW` part is communicating the merge semantics the register is using. This is similar to a "Enable-Wins Flag" CRDT type. The "type" is a `Flag` but the merge semantics is `Enable-Wins`, Same of a `Last-Write-Wins Register`.

A `Counter` is a entirely different type than a Register, as it has a richer mutation and merge mechancs. There is no additional merge semantics that need to be defined like "LWW" or "Enable-Wins" because addition is Communicative, so there doesn't need to be an additionally configured conflict strategy defined.

Therefore a `Counter` is not a `Register`

Reply:
Can you share where you get that definition of a register? When I search for LWW crdt types I can't find it associated with register and when I search for the definition of register it isn't what you say it is.